packages/flutter/lib/src/painting/image_cache.dart

```diff
@@ -132,6 +136,36 @@ class ImageCache {
     return false;
   }
 
+  /// Returns a [Future] that waits for all currently pending images to finish
+  /// loading at least one frame.
+  ///
+  /// In a Flutter widget test, this must be used with [WidgetTester.runAsync]
+  /// or [TestWidgetsFlutterBinding.runAsync].
+  @visibleForTesting
+  Future<void> waitForPendingImages() {
+    final List<Future<void>> waitList = <Future<void>>[];
+    for (final _PendingImage pendingImage in _pendingImages.values) {
+      final Completer<void> completer = Completer<void>();
+      ImageStreamListener listener;
+
+      void onDone() {
+        assert(listener != null);
+        pendingImage.completer.removeListener(listener);
+
+        completer.complete();
```

Comment:
The problem is you may want to dispose of an image handle before it's even decoded - e.g. a large list of images that a user is scrolling through rapidly and you don't know when they're going to stop. If you are forced to wait for decoding, you can end up in a deadlocked scenario where you run out of memory to decode more textures (so GPU blocks you), but you can't let go of the handles you don't even need any more because you're waiting for the GPU to decode them.

Reply:
It sounds like to me the deadlock is already an issue, and it will be worse if we make it wait for frame ready?